src/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs

```diff
@@ -43,45 +44,70 @@ public static class HostWriter
                 {
-                    BinaryUtils.SearchAndReplace(accessor, AppBinaryPathPlaceholderSearchValue, bytesToWrite);
+                    using (MemoryMappedViewAccessor accessor = memoryMappedFile.CreateViewAccessor())
+                    {
+                        BinaryUtils.SearchAndReplace(accessor, AppBinaryPathPlaceholderSearchValue, bytesToWrite);
 
-                    appHostIsPEImage = BinaryUtils.IsPEImage(accessor);
+                        appHostIsPEImage = BinaryUtils.IsPEImage(accessor);
 
-                    if (windowsGraphicalUserInterface)
-                    {
-                        if (!appHostIsPEImage)
+                        if (windowsGraphicalUserInterface)
                         {
-                            throw new AppHostNotPEFileException();
+                            if (!appHostIsPEImage)
+                            {
+                                throw new AppHostNotPEFileException();
+                            }
+
+                            BinaryUtils.SetWindowsGraphicalUserInterfaceBit(accessor);
                         }
+                    }
+                }
 
-                        BinaryUtils.SetWindowsGraphicalUserInterfaceBit(accessor);
+                if (assemblyToCopyResorcesFrom != null && appHostIsPEImage)
+                {
+                    if (ResourceUpdater.IsSupportedOS())
+                    {
+                        // Copy resources from managed dll to the apphost
+                        new ResourceUpdater(appHostDestinationFilePath)
+                            .AddResourcesFromPEImage(assemblyToCopyResorcesFrom)
+                            .Update();
+                    }
+                    else
+                    {
+                        throw new AppHostCustomizationUnsupportedOSException();
                     }
                 }
-            }
 
-            if (assemblyToCopyResorcesFrom != null && appHostIsPEImage)
+                // Memory-mapped write does not updating last write time
+                File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
+            }
+            catch (Exception ex)
             {
-                if (ResourceUpdater.IsSupportedOS())
+                FailedToDeleteApphostException failedToDeleteApphostException = null;
+                // Delete the destination file so we don't leave an unmodified apphost
+                try
                 {
-                    // Copy resources from managed dll to the apphost
-                    new ResourceUpdater(appHostDestinationFilePath)
-                        .AddResourcesFromPEImage(assemblyToCopyResorcesFrom)
-                        .Update();
+                    File.Delete(appHostDestinationFilePath);
                 }
-                else 
+                catch (Exception failedToDeleteEx) when (failedToDeleteEx is IOException || failedToDeleteEx is UnauthorizedAccessException)
```

Comment:
@peterhuene @nguerrera this is the only non straight forward copy-paste compare with https://github.com/dotnet/sdk/pull/3427/files#diff-b3b03fee742dfb740c837b74d3d814bbR104

Since this library does not have a logger. The only way I think of to surface 2 errors (first failed to embed and then failed to delete the incomplete apphost) is to use AggregateException